cabal-install/Distribution/Client/DistDirLayout.hs

```diff
@@ -107,14 +111,20 @@ data DistDirLayout = DistDirLayout {
      }
 
 
+data StoreDirLayout = StoreDirLayout {
```

Comment:
> The strategy suggested here is similar to what chak described at https://ghc.haskell.org/trac/ghc/ticket/11587#comment:4 (which you have commented in.)

Yes, this is indeed the strategy I was thinking of for OSX.

It's true that for ELF we cannot include the dir names. I wonder what nix does. At least on ELF we don't have stupid limitations, the downside of a poor scheme is just that it slows down the loading process.